sklearn/metrics/pairwise.py

```diff
@@ -917,20 +932,27 @@ def gower_distances(X, Y=None, categorical_features=None, scale=True):
         if X.shape[0] != Y.shape[0] or X is not Y:
             j_start = 0
 
-        # Calculates the similarities for categorical columns
-        cat_dists = (X[i, cat_mask] != Y[j_start:, cat_mask]) | \
-                    (_object_dtype_isnan(X[i, cat_mask]) |
-                     _object_dtype_isnan(Y[j_start:, cat_mask]))
+        # Makes the comparisson for np.nan for arrays with dtype=np.object,
+        # this is necessary as some deployments returns True for
+        # np.nan == np.nan
+        cat_nan_cols = (_object_dtype_isnan(X[i, cat_mask]) |
+                        _object_dtype_isnan(Y[j_start:, cat_mask]))
 
+        # Calculates the similarities for categorical columns
+        cat_dists = ((X[i, cat_mask] != Y[j_start:, cat_mask]) | cat_nan_cols)
         # Calculates the Manhattan distances for numerical columns
-        num_dists = abs(X[i, num_mask].astype(np.float32) -
-                        Y[j_start:, num_mask].astype(np.float32)) / ranges
+        num_dists = abs(X[i, num_mask] -
+                        Y[j_start:, num_mask]) / ranges
+
         # Calculates the number of non missing columns
-        n_missing = X.shape[1] - (np.isnan(cat_dists).sum(axis=1) +
-                                  np.isnan(num_dists).sum(axis=1))
+        n_missing = X.shape[1] - (cat_nan_cols.sum(axis=1) +
+                                  _object_dtype_isnan(num_dists).sum(axis=1))
         # Gets the final results
-        results = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
-        results /= n_missing
+        sum = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
+
+        results = np.divide(sum, n_missing,
+                            out=np.full(sum.shape, np.nan, dtype=np.object),
```

Comment:
https://github.com/scikit-learn/scikit-learn/pull/9555#discussion_r316609898
In the previous review I was asked why the cast to float32, then I removed the cast, and now all the arrays are np.object. Without this conversion we'll get an type error, because the alternative array is from another dtype than the sum array.

Reply:
This is the error I get when I don't convert the output array to np.object:

```
            results = np.divide(sum, non_missing,
                                out=np.full(sum.shape, np.nan),
>                               where=non_missing != np.zeros(sum.shape))
E           TypeError: ufunc 'true_divide' output (typecode 'O') could not be coerced to provided output parameter (typecode 'd') according to the casting rule ''same_kind''

sklearn\metrics\pairwise.py:955: TypeError
```